flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricQueryService.java

```diff
@@ -136,20 +134,53 @@ public void unregister(Metric metric) {
 	}
 
 	private void unregisterOversizedMetrics(MetricDumpSerialization.MetricSerializationResult serializationResult) {
-		for (Metric metric : serializationResult.counters.keySet()) {
-			unregister(metric);
+		if (serializationResult.serializedCounters.length > maximumFramesize) {
+			LOG.warn("The serialized counter metric is larger than the maximum frame size, " +
+				" so maybe not all metrics would be reported.");
+			serializationResult.serializedCounters = new byte[0];
```

Comment:
instead of mutating the result you can create a new one since the constructor is public.

Reply:
@zentol updated and added the test case.